File: contest/ICPC/2014/欧洲/NWERC西北欧/all/hyacinth/submissions/accepted/tobi_adjmatrix.java
import java.util.*;
import java.io.*;

public class tobi_adjmatrix {
	public static int N, from, to, channel[][] = new int[10005][2];
	public static boolean adj[][] = new boolean[10005][10005];
	public static LinkedList<Integer> nodeQ = new LinkedList<>();
	public static LinkedList<Integer> freqQ = new LinkedList<>();
	public static int numChan = 42;
	public static boolean visited[] = new boolean[10005];

	public static void main(String args[]) {
		Scanner sc = new Scanner(System.in);
		N = sc.nextInt();
		if (N == 2) {
			System.out.println("0 1\n0 1");
			return;
		}
		
		for (int i = 0; i < N - 1; i++) {
			from = sc.nextInt() - 1;
			to = sc.nextInt() - 1;
			adj[from][to] = adj[to][from] = true;
		}
		for (int i = 0; i < N; i++) {
			int cnt = 0;
			for (int j = 0; j < N; j++) {
				if (adj[i][j]) cnt++;
			}
			if (cnt == 1) {
				nodeQ.add(i);
				freqQ.add(numChan++);
				break;
			}
		}
		while (nodeQ.size() > 0) {
			int node = nodeQ.poll(), freq = freqQ.poll();
			if (visited[node]) continue;
			visited[node] = true;
			channel[node][0] = freq;
			channel[node][1] = numChan++;
			for (int i = 0; i < N; i++) if (adj[node][i] && !visited[i]) {
				freqQ.add(channel[node][1]);
				nodeQ.add(i);
			}
		}
		for (int i = 0; i < N; i++) System.out.println(channel[i][0] + " " + channel[i][1]);
	}
}

File: contest/ICPC/2014/欧洲/NWERC西北欧/all/hyacinth/submissions/accepted/tobi_adjmatrix.cpp
#include <iostream>
#include <vector>
#include <map>
#include <deque>

using namespace std;
int N, from, to, channel[10005][2];
bool adj[10005][10005];
map<int,bool> usedChannel;
bool visited[10005];
deque<int> nodeQ, freqQ;

int randChannel() {
	int c;
	do {
		c = rand() % 1000000000;
	} while (usedChannel[c]);
	usedChannel[c] = true;
	return c;
}

int main() {
	srand(time(NULL));
	cin >> N;
	if (N == 2) {
		channel[0][0] = randChannel();
		channel[0][1] = randChannel();
		cout << channel[0][0] << " " << channel[0][1] << endl;
		cout << channel[0][0] << " " << channel[0][1] << endl;
		return 0;
	}
	for (int i = 0; i < N - 1; i++) {
		cin >> from >> to;
		from--, to--;
		adj[from][to] = adj[to][from] = true;
	}
	for (int i = 0; i < N; i++) {
		int cnt = 0;
		for (int j = 0; j < N; j++) {
			if (adj[i][j]) cnt++;
		}
		if (cnt == 1) {
			nodeQ.push_back(i);
			break;
		}
	}
	freqQ.push_back(randChannel());
	while (nodeQ.size() > 0) {
		int node = nodeQ[0], freq = freqQ[0];
		nodeQ.pop_front(); freqQ.pop_front();
		if (visited[node]) continue;
		visited[node] = true;
		int which = rand() % 2;
		channel[node][which] = freq;
		channel[node][1-which] = randChannel();
		for (int i = 0; i < N; i++) if (adj[node][i] && !visited[i]) {
			freqQ.push_back(channel[node][1-which]);
			nodeQ.push_back(i);
		}
	}
	for (int i = 0; i < N; i++) {
		cout << channel[i][0] << " " << channel[i][1] << endl;
	}

	return 0;
}
